ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3MultipartCompleteUploadHandler.java

```
@@ -415,4 +430,42 @@ List<ChunkInfo> getChunksToStitch(CompleteMultipartUpload completeMultipartUploa
       return chunkInfos;
     }
   }
+
+  /**
+   * Check the list size and part number before processing request
+   * 1. Disallow duplicate part numbers
+   * 2. Disallow duplicate etags
+   * 3. Check for list size 10000
+   * 4. Check for part numbers integer 1-10000
+   * @param parts sorted parts list
+   * @return the bad request error
+   */
+  private static String validateParts(List<Part> parts) {
+    if (parts.size() > MAX_LIST_SIZE) {
+      return String.format("Parts list size cannot exceed {}.", MAX_LIST_SIZE);
+    }
+
+    Set<Integer> partNumbers = new HashSet<>();
+    Set<String> etags = new HashSet<>();
+
+    for (Part part : parts) {
+      int partNumber = part.getPartNumber();
+      String etag = part.geteTag();
+
+      if (partNumber < MIN_PART_NUM || partNumber > MAX_PART_NUM) {
+        return String.format(
+            "Invalid part number: " + part.getPartNumber() + ". Part number must be an integer between {} and {}.",
+            MIN_PART_NUM, MAX_PART_NUM);
+      }
+
+      if (!partNumbers.add(partNumber)) {
+        return "Duplicate part number found: " + partNumber;
```

Throw an exception instead of return a string.
```
          throw new RestServiceException(message, RestServiceErrorCode.BadRequest);

```